kong/core/plugins_iterator.lua

```diff
@@ -49,6 +46,64 @@ local function load_plugin_configuration(api_id, consumer_id, plugin_name)
   end
 end
 
+local function get_next(self)
+  local i = self.i
+  i = i + 1
+
+  local plugin = self.loaded_plugins[i]
+  if not plugin then
+    return nil
+  end
+
+  self.i = i
+
+  local ctx = self.ctx
+
+  -- load the plugin configuration in early phases
+  if self.access_or_cert_ctx then
+    local api = self.api
+    local plugin_configuration
+
+    local consumer = ctx.authenticated_consumer
+    if consumer then
+      local consumer_id = consumer.id
+      local schema      = plugin.schema
+
+      if not schema.no_consumer then
+        if api then
+          plugin_configuration = load_plugin_configuration(api.id, consumer_id, plugin.name)
+        end
+        if not plugin_configuration then
+          plugin_configuration = load_plugin_configuration(nil, consumer_id, plugin.name)
+        end
+      end
+    end
+
+    if not plugin_configuration then
+      -- Search API specific, or global
+      if api then
+        plugin_configuration = load_plugin_configuration(api.id, nil, plugin.name)
+      end
+      if not plugin_configuration then
+        plugin_configuration = load_plugin_configuration(nil, nil, plugin.name)
+      end
+    end
+
+    ctx.plugins_for_request[plugin.name] = plugin_configuration
+  end
+
+  -- return the plugin configuration
+  local plugins_for_request = ctx.plugins_for_request
+  if plugins_for_request[plugin.name] then
+    return plugin, plugins_for_request[plugin.name]
+  end
+
+  return get_next(self) -- Load next plugin
+end
+
+local plugin_iter_mt = { __call = get_next }
+local plugin_iter_state = setmetatable({}, plugin_iter_mt)
```

Comment:
This makes me nervous. `plugin_iter_state` is a piece of state which now exists. Previously it didn't. I am not 100% sure this won't create concurrency edge cases. I would be more confortable if a new table was returned by `iter_plugins_for_req`. There might be some loss of efficiency.